### PR TITLE
IPP code refactored.

### PR DESCRIPTION
**Pull request by dbudniko**
IPP code path  implemented as separate static functions to simplify future work with IPP code and make it more readable.

```
check_regression=*
```

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
looks reasonable :+1: 

please, fix compilation errors http://pullrequest.opencv.org/buildbot/builders/precommit_macosx/builds/2933/steps/compile%20release/logs/warnings%20%289%29

File: modules/core/src/precomp.hpp [hunk 1]
```diff
@@ -72,6 +72,20 @@
 #define GET_OPTIMIZED(func) (func)
 #endif
 
+inline void secure_cpy(void *dst,
```
Comment (alalek) [inline modules/core/src/precomp.hpp]: Where is it used?

File: modules/imgproc/src/thresh.cpp [hunk 2]
```diff
@@ -904,7 +904,34 @@ thresh_32f( const Mat& _src, Mat& _dst, float thresh, float maxval, int type )
     }
 }
 
+#ifdef HAVE_IPP
+#ifdef _MSC_VER
+#pragma warning( push )
+#pragma warning( disable : 4100 )
+#endif
```
Comment (alalek) [inline modules/imgproc/src/thresh.cpp]: May be introduce and use macro something like this:
```
#define CV_UNUSED(name) (void)name
```

**Comment by dbudniko** [comment]
Finally I've got all builds without warnings and errors. What will be the next step?

**Comment by vpisarev** [comment]
@dbudniko, thanks for the patch! I do not understand some of the transformations though. There is no code size reductions in most cases, there are extra calls added. Eventually we plan to move low-level functions from core/imgproc to the newly introduced opencv_hal, but in this case some of the transformations need to be undone, some need to be made in a different way etc. So, unless there is some strong reason[s] behind such transformations, I would put it off for now

**Comment by alalek** [comment]
This patch move out most part of optimized code from the main function code. Now IPP code looks like OpenCL code and I believe plain implementation should become more readable.

**Comment by vpisarev** [comment]
in some cases - maybe. in other cases like add8u => {add8u => ipp_add8u} I do not see any advantage of the new scheme, just increased overhead.

**Comment by lupustr3** [comment]
We wanted to use single code style for IPP calls and single call point in functions. I don't see any significant overhead of this approach, but we sure can omit such style for small functions if you like.

Overall these changes can be very helpful for IPP support and integration.

**Comment by vpisarev** [comment]
@dbudniko, @lupustr3, thank you! now it looks better. I'd postpone though integration of this patch till after 3.0. The patch is quite big and touches many basic functions. So, let's play safe.

**Comment by lupustr3** [comment]
@vpisarev, Gold was released, congratulations! )
Can we reopen this discussion?

File: modules/core/src/stat.cpp [hunk 3]
```diff
@@ -1291,51 +1307,62 @@ static bool ocl_countNonZero( InputArray _src, int & res )
...
+    IppiSize roiSize = { src.cols, src.rows };
+    Ipp32s srcstep = (Ipp32s)src.step;
+    if (src.isContinuous())
+    {
+        roiSize.width = (Ipp32s)src.total();
+        roiSize.height = 1;
+        srcstep = (Ipp32s)src.total() * CV_ELEM_SIZE(type);
+    }
+
+    if (depth == CV_8U)
```
Comment (alalek) [inline modules/core/src/stat.cpp]: This code doesn't compile with IPP (not IPPICV): `depth` is not defined here

**Comment by lupustr3** [comment]
Since Dmitry is unavailable and cannot push changes, I created new pull-request
https://github.com/Itseez/opencv/pull/4135

**Pull request closed without merging**